### ISSUE TITLE
Dev wheel upload is broken (actually it uploaded but then get deleted)

### ISSUE THREAD

**Issue by pllim** [astropy/astropy#15935]
- [ ] Revert https://github.com/astropy/astropy/commit/459e5e38287772e674c87eb3706563fa0ce97f03

As @larrybradley reported, grabbing astropy dev wheel from https://pypi.anaconda.org/astropy/simple/astropy/ now gets you a very old `dev98` wheel. I think OpenAstronomy workflow is somehow deleting the new wheels built in the nightly job instead of uploading it. See log at https://github.com/astropy/astropy/actions/runs/7620855779/job/20756741326

I don't know if https://github.com/OpenAstronomy/github-actions-workflows/pull/178 is going to fix it or not, but something needs fixing fast as this breaks all downstream  testing against astropy-dev.

https://github.com/astropy/astropy/blob/7759561c8d8760e657cee18c22636ed2ef9fd40a/.github/workflows/publish.yml#L35

cc @ConorMacBride @astrofrog @Cadair 

### Blocked by

* https://github.com/OpenAstronomy/publish-wheels-anaconda/issues/4

**Comment by pllim** [comment]
I don't grok what https://github.com/OpenAstronomy/publish-wheels-anaconda/blob/main/remove_old_wheels.py is doing but maybe the real bug is here.

**Comment by pllim** [comment]
Trying temporary workaround: https://github.com/astropy/astropy/commit/459e5e38287772e674c87eb3706563fa0ce97f03

This isn't sustainable because if we don't prune, the disk quota will kick in at some point.

**Comment by larrybradley** [comment]
Indeed the newly-build wheels get promptly deleted by `remove_old_wheels`:
https://github.com/astropy/astropy/actions/runs/7620855779/job/20756741326#step:5:2306

**Comment by larrybradley** [comment]
`remove_old_wheels` needs better sorting logic for comparing 2- and 3-digit dev versions:
https://github.com/OpenAstronomy/publish-wheels-anaconda/blob/main/remove_old_wheels.py#L26

**Comment by ConorMacBride** [comment]
Yes, the current sorting is alphabetical on the version string. It looks like we need a fix for this issue: https://github.com/OpenAstronomy/publish-wheels-anaconda/issues/4

**Comment by ConorMacBride** [comment]
The sorting issue has been fixed, so you should be able to enable pruning again.

**Comment by pllim** [comment]
Thanks!

**Comment by pllim** [comment]
I re-enabled the pruning at 0097d6a08c479402d5d6fd1e4487bcf35584f99d . @larrybradley , please open new issue if you see something weird again. Given the currently failures caused by pytest 8, we might not know for sure until that is resolved. Thanks, all!

**Comment by pllim** [comment]
@ConorMacBride , does the listing now look right to you (from https://github.com/astropy/astropy/actions/runs/7717116721/job/21041198528)? https://pypi.anaconda.org/astropy/simple/astropy/

I see the latest wheel uploaded (dev389) but I also see very old ones staying (dev78, dev88, dev90, dev94, dev98).

**Comment by ConorMacBride** [comment]
Very strange. Looks like the sorting isn't being applied, but I can't see any bugs in the code. I definitely tested the script on the astropy conda repository and it selected the right versions. I'll generate a new API key later this week and test again. Maybe just disable pruning again, sorry.

**Comment by pllim** [comment]
No worries. The latest wheel uploaded, so it's not a fire. Just weird that the very old wheels won't go away. Do you want me to open a new issue?

**Comment by bsipocz** [comment]
@pllim - fwiw, have a look at the scientific python action, it's well maintained and is used by a lot of other libraries, and is now capable to upload to a different anaconda space (pruning comes out of the box).

**Comment by ConorMacBride** [comment]
@pllim actually it looks to be working property. I misread the logs before. The previous alphabetical sorting meant that there were already big gaps between the build numbers currently on anaconda. Based on an API response I had saved it would have calculated the following when running after the bug fix:

- `6.1.dev48+gbe40cc47de` [DELETE]
- `6.1.dev56+ge81466f73f` [KEEP]
- `6.1.dev59+g1ae26ae0fc` [KEEP]
- `6.1.dev69+g83fcda46f8` [KEEP]
- `6.1.dev78+gbc80072326` [KEEP]
- `6.1.dev88+gc5b3fe5b4a` [KEEP]
- `6.1.dev90+g5f05c10d35` [KEEP]
- `6.1.dev94+g1e634ee1c4` [KEEP]
- `6.1.dev98+gaa0b136c4d` [KEEP]
- `6.1.dev359+g459e5e3828` [KEEP]
- `6.1.dev1000+g459e5e3828` [KEEP] (new upload)

Looking at more recent logs the new build are not being deleted and the oldest ones are starting to disappear.

**Comment by pllim** [comment]
I thought you changed it to date based, so theoretically it would have deleted very old wheels?

**Comment by ConorMacBride** [comment]
The sorting is date based, but it still keeps the N latest versions uploaded. Due to the incorrect sorting previously more recent builds had already been deleted so the 10th most recent was actually quite old.
